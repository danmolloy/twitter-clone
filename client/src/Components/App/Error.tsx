import { GlobeIcon, RefreshIcon } from "@heroicons/react/outline"

export const Error = () => {

  const handleClick = () => {
    alert(Object.keys(localStorage))
    window.location.reload()
  }

  return (
    <div id="error-page" className="w-full flex flex-col items-center mt-12">
      <GlobeIcon className="h-6 w-auto"/>
      <p className="text-gray-600">An error occurred. Please check your connection and try again.</p>
      <button className="tweet-btn py-1 flex flex-row" onClick={() => handleClick()}>
        <RefreshIcon className="h-6 w-auto"/>
        <p className="font-semibold px-1">Retry</p>
      </button>
    </div>
  )
}