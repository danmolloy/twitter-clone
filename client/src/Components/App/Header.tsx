export const Header = (props: any) => {
  return (
    <div className="w-full p-2 border-b border-r">
      <h2 className="text-xl font-semibold">{props.pageTitle}</h2>
      {props.blurb &&
      <p>{props.blurb}</p>}
    </div>
  )
}