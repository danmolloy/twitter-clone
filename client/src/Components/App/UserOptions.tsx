import { LogoutIcon, TrashIcon, UserIcon, XCircleIcon } from "@heroicons/react/outline"
import { Link } from "react-router-dom"

export const UserOptions = (props: any) => {

  
  return (
    <div className="flex flex-col items-start fixed bg-white mb-40 -ml-20 border z-20">
      <button onClick={() => props.close()} className="self-end" title="Hide Menu">
          <XCircleIcon className="w-10 p-1 h-auto text-red-500 hover:bg-red-50 rounded-full mr-2"/>
        </button>
      <Link to={`/${props.currentUser.handle.slice(1)}`} title="Profile" className="lg-side-icon">
        <UserIcon className="side-icon flex"/>
      </Link>
      <Link title="Sign out" className="lg-side-icon" to="/home" onClick={() => props.onSignOut()}>
        <LogoutIcon className="side-icon flex" id="more-icon"/>
      </Link>
      <Link title="Delete user" to="/home" className="lg-side-icon hover:bg-red-50" onClick={() => props.deleteUser()}>
        <TrashIcon id="delete-icon" className="side-icon flex text-red-500 hover:bg-red-50"/>
      </Link>
    </div>
  )
}