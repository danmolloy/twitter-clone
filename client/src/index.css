@tailwind base;
@tailwind components;
@tailwind utilities;

.side-icon {
  @apply h-12 w-auto mx-8 sm:mx-2 rounded-full p-2 hover:bg-gray-200;
}

.lg-side-icon {
  @apply hover:bg-gray-200 flex flex-row items-center md:ml-6 md:pr-4 my-1 rounded-full;
}

.sidebar-text {
  @apply hidden md:flex text-lg;
}

.compose-tweet-icons {
  @apply w-10 h-auto twitter-blue p-2 hover:bg-blue-50 rounded-full;
}

.tweet-options {
  @apply w-10 h-auto p-2 rounded-full;
}

.tweet-btn {
  @apply h-8 bg-twitter-blue text-white px-3 my-3 rounded-full font-bold mr-3 shadow;
}

.tweet-btn-disabled {
  @apply h-8 bg-blue-opacity text-white px-3 my-3 rounded-full font-bold mr-3;
}

.r-bar-article {
  @apply p-2 w-full;
}

.welcome-background {
  background-image: linear-gradient(to bottom right, #ffffff, #1da1f2);
}


.blue-hover {
  background-color: #1DA1F2;
  opacity: .5;
}

.bg-twitter-blue {
  background-color: #1DA1F2;
}

.twitter-blue {
  --tw-text-opacity: 1;
  color: #1DA1F2;
}


.bg-blue-opacity {
  background-color: #1DA1F2;
  opacity: .5;
}


svg path {
  stroke-width: 1.5;
}

.side-bar {
  @apply self-end fixed flex flex-row sm:flex-col sm:mt-0 ml-0 w-full 
  sm:w-24 md:w-60 sm:h-screen border-t sm:border-t-0 sm:border-r 
  border-gray-200 md:items-start justify-center sm:justify-start items-center bg-white;
}

/* Home */
.home-header {
  @apply border-b border-gray-200 h-14 flex flex-row justify-between;
}

.h2 {
  @apply text-xl font-semibold;
}

/* Notifications */
.notification-btn-selected {
  @apply font-semibold w-1/2 h-12 hover:bg-gray-200 flex flex-col items-center justify-between pt-2;
}

.notification-btn {
  @apply  w-1/2 h-12 hover:bg-gray-200 flex flex-col items-center justify-between pt-2 text-gray-500;
}

.tab-line {
  @apply border bg-twitter-blue h-1.5 rounded-full w-12;
}

/* Profile*/

.selected-tweet-filter {
  @apply w-1/2 border-b font-semibold hover:bg-gray-200 flex flex-col justify-between items-center pt-2;
}

.deselected-tweet-filter {
  @apply w-1/2 hover:bg-gray-200 border-b text-center py-2;
}

.follow-btn {
  @apply font-bold bg-black text-white rounded-full py-2 px-4 my-8 mr-8;
}

.unfollow-btn {
  @apply font-bold border border-gray-300 rounded-full py-2 px-4 my-8 mr-8;
}

textarea {
  resize: none
}