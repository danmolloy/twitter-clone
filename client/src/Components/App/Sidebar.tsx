import { HashtagIcon, 
  HomeIcon, 
  UserIcon, 
  BellIcon,
  InboxIcon,
  LogoutIcon,
} from '@heroicons/react/outline'
import { Link } from 'react-router-dom'
import { AUTH_TOKEN } from '../../constants'
import { gql, useQuery } from '@apollo/client'
import { Error } from './Error'
import { Loading } from './Loading'
import { useState } from 'react'
import { UserOptions } from './UserOptions'
import { Message, Notification } from '../../types'

export const DELETE_USER = gql`
  mutation SignUpMutation {
    deleteUser {
      name
    }
  }
`;

export const GET_USER = gql`
  query {
    getNotifications {
      notifications {
          read
        }
        chats {
          content {
            read
            authorHandle
          }
        }
        handle
        profilePic
    }
  }
`;


export const Sidebar = () => {
  const [showUserOptions, setShowUserOptions] = useState(false)

  const { loading, error, data } = useQuery(GET_USER)

  const onSignOut = () => {
    localStorage.removeItem(AUTH_TOKEN)
    window.location.reload()
  }

  if (error) {
    console.log(error)
    return <Error />
  }

  if (loading) {
    return <Loading />
  }

  return (
    <div id="side-bar" 
    className="side-bar">
      <Link to="/home" className="md:ml-2" id="twitter-home-link" title="Home">
        <img src="/favico.ico" alt="Blue Twitter bird" className="hidden sm:flex w-12 h-auto mx-6 my-4 p-2 rounded hover:bg-blue-50"/>
      </Link>
      <Link to="/home" className="lg-side-icon" id="home-link" title="Home">
        <HomeIcon className="side-icon"/>
        <label htmlFor="home-link" className="sidebar-text">Home</label>
      </Link>
      <Link to="/explore" title="Explore" className="lg-side-icon" id="explore-link">
        <HashtagIcon className="side-icon"/>
        <label htmlFor="explore-link" className="sidebar-text">Explore</label>
      </Link>
      <Link to="/notifications" title="Notifications" className="lg-side-icon" id="notifications-link">
      {data && data.getNotifications.notifications.map(
        (i: Notification) => i.read).includes(false) 
      && <div id="notifications-alert" className="bg-blue-600 h-2 w-2 rounded-full z-10 -mr-8 sm:-mr-2 -mt-6 sm:-mt-0" />}
        <BellIcon className="side-icon"/>
        <label htmlFor="notifications-link" className="sidebar-text">Notifications</label>
      </Link>
      <Link to="/messages" title="Messages" className="lg-side-icon">
      {data && data.getNotifications.chats[0] !== undefined
      && data.getNotifications.chats[0].content.filter(
        (i:Message) => i.read === false && i.authorHandle !== data.getNotifications.handle).length > 0
        && <div id="unread-msg-alert" className="bg-blue-600 h-2 w-2 rounded-full z-10 -mr-8 sm:-mr-2 -mt-6 sm:-mt-0" />}
        <InboxIcon className="side-icon"/>
        <label htmlFor="messages-link" className="sidebar-text">Messages</label>
      </Link>
      <button className="sm:hidden lg-side-icon" title="User options" onClick={() => setShowUserOptions(!showUserOptions)} >
        {showUserOptions 
        && <UserOptions 
        currentUser={data && data.getNotifications} 
        onSignOut={() => onSignOut()} 
        close={() => setShowUserOptions(false)}/>}
        <img alt={data.getNotifications?.name ? data.getNotifications?.name : "Current user"} src={data && data.getNotifications?.profilePic} className="side-icon rounded-full"/>
      </button>
      <Link id="profile-link" to={data ? `/${data.getNotifications.handle.slice(1)}`: '/'} title="Profile" className="hidden sm:flex lg-side-icon">
        <UserIcon className="side-icon hidden sm:flex"/>
        <label htmlFor="profile-link" className="sidebar-text">Profile</label>
      </Link>
      <Link title="Sign out" id="sign-out-btn" className="hidden sm:flex lg-side-icon" to="/home" onClick={() => onSignOut()}>
        <LogoutIcon className="side-icon hidden sm:flex" id="more-icon"/>
        <label htmlFor="sign-out" className="sidebar-text">Sign Out</label>
      </Link>
    </div>
  )
}