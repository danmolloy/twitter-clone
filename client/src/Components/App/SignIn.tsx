import { useState } from "react"
import { useQuery, useMutation, gql } from '@apollo/client'
import { AUTH_TOKEN } from "../../constants";
import { UserHandles } from "../../types";


const USER_HANDLES = gql`
  query GetAllHandles {
    getAllHandles {
      handle
    }
  }
`;

const LOGIN = gql`
  mutation Mutation($handle: String!, $password: String!) {
    login(handle: $handle, password: $password) {
      token
      user {
        handle
      }
    }
  }
`;

const SIGNUP = gql`
  mutation Mutation($handle: String!, $password: String!, $name: String!) {
    signUp(handle: $handle, password: $password, name: $name) {
      token
      user {
        handle
        name
      }
    }
  }
`;

const regExFullName = new RegExp(/^[a-z]{2,}(\s)[a-z]{2,}$/i);
const regexUserName = new RegExp(/^[a-zA-Z0-9_]{3,}$/gi);

export const SignIn = () => {
  const [signUpForm, setSignUpForm] = useState(true)
  const [name, setName] = useState("")
  const [handle, setHandle] = useState("")
  const [password, setPassword] = useState("")
  const [showPassword, setShowPassword] = useState(false)
  const {loading, error, data} = useQuery(USER_HANDLES)

  const [login] = useMutation(LOGIN, { 
    variables: { 
      handle: `@${handle}`, 
      password: password
    },
    onCompleted: ({ login }) => {
      localStorage.setItem(AUTH_TOKEN, login.token);
      window.location.reload()
    }
  })

  const [signup] = useMutation(SIGNUP, { 
    variables: { 
      name: name,
      handle: `@${handle}`, 
      password: password
    },
    onCompleted: ({ signUp }) => {
      localStorage.setItem(AUTH_TOKEN, signUp.token);
      window.location.reload();
    }
  })

  const ValidateSignUp = (fullName: string, userName: string) => {
    if (data && data.getAllHandles.map((i: UserHandles) => 
      i.handle.slice(1)).includes(userName) ) {
      return alert("Username is already taken. Please try another name.")
      }

    if (!regExFullName.test(fullName)) {
      return alert("Full Name invalid. Ensure exactly two names are at least 2 characters in length, contain letters only and are seperated by a space.")
    }
    if (!regexUserName.test(userName)) {
      return alert("Username invalid. May only contain letters, numbers, exactly 0 or 1 underscores and must be between 3-12 characters long.")
    }
    else {
      signup()
    }
  }

  const checkUser = () => {
    if (data && data.getAllHandles.map((i: UserHandles) => 
    i.handle.slice(1)).includes(handle)) {
      login()
    }
    else {
      alert("User not found.")
    }
  }

  return (
    <div className="welcome-background flex flex-col items-center border fixed w-full h-full">
      <div className="sm:hidden flex flex-row items-center">
      <img alt="White Twitter bird" src="/whiteTwitterLogo.png" className="w-8 h-auto"/>
      </div>
      <div className="flex rounded shadow-md flex-col sm:self-start sm:ml-8 sm:ml-24 items-center sm:mt-8 border bg-white z-10">
      <h2 className="text-lg font-semibold">{signUpForm ? "Sign Up" : "Sign In"}</h2>
      <div className="flex flex-col p-2 max-w-xs">
          {signUpForm &&
            <div className="flex flex-col py-4">
              <label htmlFor="name-input">
                Full Name
              </label>
              <input value={name} placeholder="Dan Molloy" maxLength={20} onChange={e => setName(e.target.value)} className="px-2 border m-2 rounded-full w-72" id="name-input"/>
              <p className="text-xs text-gray-500 px-2">Must contain letters and space only</p>
            </div>}
          <div className="flex flex-col py-4">
            <label htmlFor="username-input">
              Username
            </label>
          <input placeholder="dan_molloy1" value={handle} onChange={e => setHandle(e.target.value)} maxLength={12} className="border m-2 rounded-full px-2 w-72" id="handle-input"/>
          {signUpForm && <p className="text-xs text-gray-500 px-2">Must only contain letters, numbers, underscores and must be between 3-12 characters long. Must be unique.</p>}
        </div>
        <div className="flex flex-col py-4">
          <label htmlFor="password-input">Password</label>
          <input type={showPassword ? "text" : "password"} value={password} onChange={e => setPassword(e.target.value)} maxLength={20} id="password-input" className="px-2 border m-2 rounded-full" />
          {signUpForm && <p className="text-xs text-gray-500 px-2">Must be 3-20 characters long</p>}
         <button 
          type="button" 
          onClick={() => setShowPassword(!showPassword)} className="text-sm hover:underline">{`${showPassword ? "Hide": "Show"} password`}</button>
        </div>
        <button onClick={signUpForm ? () => {
          ValidateSignUp(name, handle)}
          : () => {
            checkUser()}
          } className="tweet-btn shadow-md">Submit</button>
      </div>
      <button onClick={() => setSignUpForm(!signUpForm)} className="hover:underline twitter-blue py-2">{signUpForm ? "Already have an account?" : "Create an account"}</button>
    </div>
    <img alt="Blue Twitter bird" src="/whiteTwitterLogo.png" className="hidden sm:flex sm:fixed w-1/2 sm:w-3/4 h-auto sm:justify-end sm:ml-24 sm:ml-48 md:mt-48 lg:mt-0"/>
    </div>
  )
}