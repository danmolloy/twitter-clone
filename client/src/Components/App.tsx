import React, { useEffect, useState } from "react";
import {
  BrowserRouter as Router,
  Switch,
  Route,
  Link,
  Redirect
} from "react-router-dom"
import { Home } from './Home'
import { Explore } from './Explore'
import { Bookmarks } from './Bookmarks'
import { Messages } from './Messages'
import { Notifications } from './Notifications'
import { Profile } from './Profile'
import { Sidebar } from "./Sidebar";
import { useQuery, gql } from '@apollo/client'
import { ComposeTweet } from "./ComposeTweet";
import { RightBar }  from './RightBar'
import { Loading } from "./Loading";
import { Error } from "./Error";
import { CurrentUserData, CurrentUserVar } from '../types'
import { SignIn } from "./SignIn";
import { AUTH_TOKEN } from '../constants'
import { Chat } from "./Chat";

export const CURRENTUSER = gql`
query Query {
  currentUser {
    name
    handle
    blurb
    joinDate
    bgPic
    profilePic
    follows {
      handle
    }
    followers {
      handle
    }
    writtenPosts {
      id
      content
      postDate
      likes {
        handle
      }
    }
  }
}
`;

function App() {
  const { loading, error, data } = useQuery<CurrentUserData, CurrentUserVar>(CURRENTUSER)
  const authToken = localStorage.getItem(AUTH_TOKEN)

  if (!authToken) {
    return  <SignIn />
  }

  if (loading) {
    return <Loading />
  }

  if (error ) {
    return <Error />
  }

  return (
    <Router>
      <div id="main-content" className="flex flex-row w-screen h-screen justify-start">
        {authToken && <Sidebar currentUser={data && data.currentUser}/>}
        <div className="sm:ml-24 md:ml-60 mb-0 border-r w-full max-w-2xl sm:mr-2">
        <Switch>
          <Route path="/compose/tweet">
            <ComposeTweet currentUser={data && data.currentUser} updatePage={() => alert('Posted!')}/>
          </Route>
          <Route path="/explore">
            <Explore currentUserHandle={data && data.currentUser.handle}/>
          </Route>
          <Route path="/home">
            <Home currentUser={data && data.currentUser}/>
          </Route>
          <Route path="/bookmarks">
            <Bookmarks currentUser={data && data.currentUser} />
          </Route>
          <Route path="/messages/:chatId">
            <Chat currentUser={data && data.currentUser}/>
          </Route>
          <Route path="/messages">
            <Messages currentUser={data && data.currentUser}/>
          </Route>
          <Route exact path="/login">
            <SignIn />
          </Route>
          <Route path="/notifications" component={Notifications} />
          <Route path={'/:userHandle'}>
            <Profile currentUser={data && data.currentUser}/>
          </Route>
          <Route path="/">
            <Redirect to={authToken ? "/home" : "/login"} />
          </Route>
        </Switch>
        </div>
          <RightBar />
        </div>
    </Router>
  );
}

export default App;
