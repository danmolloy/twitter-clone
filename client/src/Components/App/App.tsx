import {
  BrowserRouter as Router,
  Switch,
  Route,
  Redirect
} from "react-router-dom"
import { Home } from '../Home/Home'
import { Explore } from '../Explore/Explore'
import { Messages } from '../Messages/Messages'
import { Notifications } from '../Notifications/Notifications'
import { Profile } from '../Profile/Profile'
import { Sidebar } from "./Sidebar";
import { useQuery, gql } from '@apollo/client'
import { ComposeTweet } from "../Home/ComposeTweet";
import { Loading } from "./Loading";
import { Error } from "./Error";
import { CurrentUserData, CurrentUserVar } from '../../types'
import { SignIn } from "./SignIn";
import { AUTH_TOKEN } from '../../constants'
import { Chat } from "../Messages/Chat";

export const CURRENTUSER = gql`
query Query {
  currentUser {
    name
    handle
    blurb
    joinDate
    bgPic
    profilePic
    chats {
      content {
        read
      }
    }
    follows {
      handle
    }
    followers {
      handle
    }
    notifications {
      id
      text
      sentFromUser
      read
      tweetId
    }
    writtenPosts {
      id
      content
      postDate
      likes {
        handle
      }
    }
  }
}
`;

function App() {
  const { loading, error, data } = useQuery<CurrentUserData, CurrentUserVar>(CURRENTUSER)
  const authToken = localStorage.getItem(AUTH_TOKEN)

  if (!authToken) {
    return  <SignIn />
  }

  if (loading) {
    return <Loading />
  }

  if (error ) {
    console.log(error)
    return <Error />
  }

  return (
    <Router>
      <div id="main-content" className="flex flex-row w-screen h-screen justify-start">
        {authToken && <Sidebar />}
        <div className="sm:ml-24 md:ml-60 mb-0 w-full max-w-2xl sm:mr-2 md:mr-0">
        <Switch>
          <Route path="/explore">
            <Explore currentUserHandle={data && data.currentUser.handle}/>
          </Route>
          <Route path="/home">
            <Home currentUser={data && data.currentUser}/>
          </Route>
          <Route path="/messages/:chatId">
            <Chat currentUser={data && data.currentUser}/>
          </Route>
          <Route path="/messages">
            <Messages currentUser={data && data.currentUser}/>
          </Route>
          <Route exact path="/login">
            <SignIn />
          </Route>
          <Route path="/notifications" component={Notifications} />
          <Route path={'/:userHandle'}>
            <Profile currentUser={data && data.currentUser}/>
          </Route>
          <Route path="/">
            <Redirect to={authToken ? "/home" : "/login"} />
          </Route>
        </Switch>
        </div>
      </div>
    </Router>
  );
}

export default App;
