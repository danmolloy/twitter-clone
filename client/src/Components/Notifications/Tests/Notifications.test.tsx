import { act } from "react-dom/test-utils";
import { unmountComponentAtNode, render } from "react-dom";
import { AUTH_TOKEN } from "../../../constants";
import { MockedProvider } from "@apollo/client/testing";
import { MemoryRouter } from "react-router";
import pretty from "pretty";
import { fireEvent } from "@testing-library/dom";
import { NotificationsMock } from "./NotificationsMock";
import App from "../../App/App";

let container: any = null


beforeEach(() => {
  container = document.createElement("div");
  document.body.appendChild(container);
  global.localStorage.setItem(AUTH_TOKEN, "auth-token");
});

afterEach(() => {
  unmountComponentAtNode(container);
  container.remove();
  container = null;
  global.localStorage.removeItem(AUTH_TOKEN)
})

describe("notifications", () => {
  it("renders page", async () => {
    act(() => {
      render(
        <MockedProvider mocks={NotificationsMock} addTypename={false}>
          <MemoryRouter initialEntries={["/notifications"]}>
            <App />
          </MemoryRouter>
        </MockedProvider>, container
      )
    })
    await act(async() => {
      await new Promise(resolve => setTimeout(resolve, 100))
    })
    expect(pretty(container.innerHTML)).toMatchSnapshot();
  })


  it("shows notification in sidebar", async() => {
    act(() => {
      render(
        <MockedProvider mocks={NotificationsMock} addTypename={false}>
          <MemoryRouter initialEntries={["/notifications"]}>
            <App />
          </MemoryRouter>
        </MockedProvider>, container
      )
    })
    await act(async() => {
      await new Promise(resolve => setTimeout(resolve, 100))
    })
    expect(pretty(container.querySelector("#side-bar").innerHTML)).toMatch(/notifications-alert/g)
  })

  it("unread is bold", async () => {
    act(() => {
      render(
        <MockedProvider mocks={NotificationsMock} addTypename={false}>
          <MemoryRouter initialEntries={["/notifications"]}>
            <App />
          </MemoryRouter>
        </MockedProvider>, container
      )
    })
    await act(async() => {
      await new Promise(resolve => setTimeout(resolve, 100))
    })
    expect(container.querySelector("#single-notification").className).toMatch(/font-semibold/g)
  })

  it("marks as read", async() => {
    act(() => {
      render(
        <MockedProvider mocks={NotificationsMock} addTypename={false}>
          <MemoryRouter initialEntries={["/notifications"]}>
            <App />
          </MemoryRouter>
        </MockedProvider>, container
      )
    })
    await act(async() => {
      await new Promise(resolve => setTimeout(resolve, 100))
    })
    expect(container.querySelector("#single-notification").className).toMatch(/font-semibold/g)
    await act(async() => {
      await new Promise(resolve => setTimeout(resolve, 1000))
    })
    expect(container.querySelector("#single-notification").className).not.toMatch(/font-semibold/g)
    expect(pretty(container.querySelector("#side-bar").innerHTML)).not.toMatch(/notifications-alert/g)

  })

  it("links to user profile", async() => {
    act(() => {
      render(
        <MockedProvider mocks={NotificationsMock} addTypename={false}>
          <MemoryRouter initialEntries={["/notifications"]}>
            <App />
          </MemoryRouter>
        </MockedProvider>, container
      )
    })
    await act(async() => {
      await new Promise(resolve => setTimeout(resolve, 100))
    })
    fireEvent.click(container.querySelector("#single-notification"))
    await act(async() => {
      await new Promise(resolve => setTimeout(resolve, 100))
    })
    expect(container.textContent).toMatch(/I call it 'Twitter'/g)
  })  
})
