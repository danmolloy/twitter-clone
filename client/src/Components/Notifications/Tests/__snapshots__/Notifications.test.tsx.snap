// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`notifications renders page 1`] = `
"<div>
  <div id=\\"main-content\\" class=\\"flex flex-row w-screen h-screen justify-start\\">
    <div id=\\"side-bar\\" class=\\"side-bar\\"><a class=\\"md:ml-2\\" id=\\"twitter-home-link\\" title=\\"Home\\" href=\\"/home\\"><img src=\\"/favico.ico\\" alt=\\"Blue Twitter bird\\" class=\\"hidden sm:flex w-12 h-auto mx-6 my-4 p-2 rounded hover:bg-blue-50\\"></a><a class=\\"lg-side-icon\\" id=\\"home-link\\" title=\\"Home\\" href=\\"/home\\"><svg xmlns=\\"http://www.w3.org/2000/svg\\" fill=\\"none\\" viewBox=\\"0 0 24 24\\" stroke=\\"currentColor\\" class=\\"side-icon\\">
          <path stroke-linecap=\\"round\\" stroke-linejoin=\\"round\\" stroke-width=\\"2\\" d=\\"M3 12l2-2m0 0l7-7 7 7M5 10v10a1 1 0 001 1h3m10-11l2 2m-2-2v10a1 1 0 01-1 1h-3m-6 0a1 1 0 001-1v-4a1 1 0 011-1h2a1 1 0 011 1v4a1 1 0 001 1m-6 0h6\\"></path>
        </svg><label for=\\"home-link\\" class=\\"sidebar-text\\">Home</label></a><a title=\\"Explore\\" class=\\"lg-side-icon\\" id=\\"explore-link\\" href=\\"/explore\\"><svg xmlns=\\"http://www.w3.org/2000/svg\\" fill=\\"none\\" viewBox=\\"0 0 24 24\\" stroke=\\"currentColor\\" class=\\"side-icon\\">
          <path stroke-linecap=\\"round\\" stroke-linejoin=\\"round\\" stroke-width=\\"2\\" d=\\"M7 20l4-16m2 16l4-16M6 9h14M4 15h14\\"></path>
        </svg><label for=\\"explore-link\\" class=\\"sidebar-text\\">Explore</label></a><a title=\\"Notifications\\" class=\\"lg-side-icon\\" id=\\"notifications-link\\" href=\\"/notifications\\">
        <div id=\\"notifications-alert\\" class=\\"bg-blue-600 h-2 w-2 rounded-full z-10 -mr-8 sm:-mr-2 -mt-6 sm:-mt-0\\"></div><svg xmlns=\\"http://www.w3.org/2000/svg\\" fill=\\"none\\" viewBox=\\"0 0 24 24\\" stroke=\\"currentColor\\" class=\\"side-icon\\">
          <path stroke-linecap=\\"round\\" stroke-linejoin=\\"round\\" stroke-width=\\"2\\" d=\\"M15 17h5l-1.405-1.405A2.032 2.032 0 0118 14.158V11a6.002 6.002 0 00-4-5.659V5a2 2 0 10-4 0v.341C7.67 6.165 6 8.388 6 11v3.159c0 .538-.214 1.055-.595 1.436L4 17h5m6 0v1a3 3 0 11-6 0v-1m6 0H9\\"></path>
        </svg><label for=\\"notifications-link\\" class=\\"sidebar-text\\">Notifications</label>
      </a><a title=\\"Messages\\" class=\\"lg-side-icon\\" href=\\"/messages\\"><svg xmlns=\\"http://www.w3.org/2000/svg\\" fill=\\"none\\" viewBox=\\"0 0 24 24\\" stroke=\\"currentColor\\" class=\\"side-icon\\">
          <path stroke-linecap=\\"round\\" stroke-linejoin=\\"round\\" stroke-width=\\"2\\" d=\\"M20 13V6a2 2 0 00-2-2H6a2 2 0 00-2 2v7m16 0v5a2 2 0 01-2 2H6a2 2 0 01-2-2v-5m16 0h-2.586a1 1 0 00-.707.293l-2.414 2.414a1 1 0 01-.707.293h-3.172a1 1 0 01-.707-.293l-2.414-2.414A1 1 0 006.586 13H4\\"></path>
        </svg><label for=\\"messages-link\\" class=\\"sidebar-text\\">Messages</label></a><button class=\\"sm:hidden lg-side-icon\\" title=\\"User options\\"><img alt=\\"Current user\\" src=\\"profilePic.jpg\\" class=\\"side-icon rounded-full\\"></button><a id=\\"profile-link\\" title=\\"Profile\\" class=\\"hidden sm:flex lg-side-icon\\" href=\\"/leGuin\\"><svg xmlns=\\"http://www.w3.org/2000/svg\\" fill=\\"none\\" viewBox=\\"0 0 24 24\\" stroke=\\"currentColor\\" class=\\"side-icon hidden sm:flex\\">
          <path stroke-linecap=\\"round\\" stroke-linejoin=\\"round\\" stroke-width=\\"2\\" d=\\"M16 7a4 4 0 11-8 0 4 4 0 018 0zM12 14a7 7 0 00-7 7h14a7 7 0 00-7-7z\\"></path>
        </svg><label for=\\"profile-link\\" class=\\"sidebar-text\\">Profile</label></a><a title=\\"Sign out\\" id=\\"sign-out-btn\\" class=\\"hidden sm:flex lg-side-icon\\" href=\\"/home\\"><svg xmlns=\\"http://www.w3.org/2000/svg\\" fill=\\"none\\" viewBox=\\"0 0 24 24\\" stroke=\\"currentColor\\" class=\\"side-icon hidden sm:flex\\" id=\\"more-icon\\">
          <path stroke-linecap=\\"round\\" stroke-linejoin=\\"round\\" stroke-width=\\"2\\" d=\\"M17 16l4-4m0 0l-4-4m4 4H7m6 4v1a3 3 0 01-3 3H6a3 3 0 01-3-3V7a3 3 0 013-3h4a3 3 0 013 3v1\\"></path>
        </svg><label for=\\"sign-out\\" class=\\"sidebar-text\\">Sign Out</label></a></div>
    <div class=\\"sm:ml-24 md:ml-60 mb-0 w-full max-w-2xl sm:mr-2 md:mr-0\\">
      <div id=\\"notifications-component\\" class=\\"w-full h-full flex flex-col\\">
        <div class=\\"flex flex-col border-b border-r\\">
          <h2 class=\\"text-xl font-semibold pl-3 pt-3 pb-3\\">Notifications</h2>
        </div>
        <div class=\\"w-full h-screen flex flex-col border-r\\"><a id=\\"single-notification\\" class=\\"flex flex-row justify-between hover:bg-gray-200 border-b w-full p-6 font-semibold bg-gray-50\\" href=\\"/danmolloy\\">
            <p>@danmolloy has started following you.</p>
            <p class=\\"text-gray-500\\">Wed Nov 24 2021</p>
          </a></div>
      </div>
    </div>
  </div>
</div>"
`;
